src/backend/distributed/utils/shardinterval_utils.c

```diff
@@ -273,10 +254,14 @@ FindShardInterval(Datum partitionColumnValue, ShardInterval **shardIntervalCache
  * fire this.
  */
 static int
-FindShardIntervalIndex(Datum searchedValue, ShardInterval **shardIntervalCache,
-					   int shardCount, char partitionMethod, FmgrInfo *compareFunction,
-					   bool useBinarySearch)
+FindShardIntervalIndex(Datum searchedValue, DistTableCacheEntry *cacheEntry)
 {
+	ShardInterval **shardIntervalCache = cacheEntry->sortedShardIntervalArray;
+	int shardCount = cacheEntry->shardIntervalArrayLength;
+	char partitionMethod = cacheEntry->partitionMethod;
+	FmgrInfo *compareFunction = cacheEntry->shardIntervalCompareFunction;
```

Comment:
I wonder if it's worth passing down the cache entry to `SearchCachedShardInterval` as well. 

(It also would be the most technically correct solution. It's very unlikely, but a compare function call could trigger a cache invalidation if it calls into some Citus logic that does a cache lookup. That would then cause the next compare function in the binary search to segfault.)